Move secrets to Azure Key Vault

mjcheetham opened the pull request:
We should standardise on a location (and naming convention) to store secrets used as part of the build process for `microsoft/git`. Azure Key Vault is an approved store for secrets inside of Microsoft, and so we should migrate any existing secrets from GitHub environment secrets to AKV. Access to the AKV is via Managed Identity and [federated authentication through GitHub Actions](https://learn.microsoft.com/en-us/azure/developer/github/connect-from-azure-openid-connect).

The names of the secrets, and other configuration for accessing the Key Vault, remain in GitHub environment secrets. Forks of the project can therefore define the same set of environment secrets pointing at their own AKV to utilise the same build process.

For reference, the set of secrets that must be defined for the workflow are as follows:

Secret Name|Description
-|-
`AZURE_VAULT`|Name of the Azure Key Vault containing build secrets.
`AZURE_CLIENT_ID`|Client ID of the Managed Identity with access to the Key Vault.
`AZURE_TENANT_ID`|Tenant ID where the Managed Identity resides.
`AZURE_SUBSCRIPTION_ID`|Subscription ID where the Key Vault resides.
`WIN_CODESIGN_CERT_SECRET_NAME`|Name of the AKV secret containing the base64 encoded Windows Authenticode signing certificate.
`WIN_CODESIGN_PASS_SECRET_NAME`|Name of the AKV secret containing the password for the Windows Authenticode signing certificate.
`WIN_GPG_PRIVATE_SECRET_NAME`|Name of the AKV secret containing the GPG private key used to sign the Windows package.
`WIN_GPG_KEYGRIP_SECRET_NAME`|Name of the AKV secret containing the keygrip for the GPG key used to sign the Windows package.
`WIN_GPG_PASSPHRASE_SECRET_NAME`|Name of the AKV secret containing the passphrase for the GPG private key used to sign the Windows package.
`APPLE_APPCERT_SECRET_NAME`|Name of the AKV secret containing the base64 encoded Apple Application Certificate for macOS.
`APPLE_APPCERT_PASS_SECRET_NAME`|Name of the AKV secret containing the password for the Apple Application Certificate.
`APPLE_INSTCERT_SECRET_NAME`|Name of the AKV secret containing the base64 encode Apple InstallationCertificate for macOS.
`APPLE_INSTCERT_PASS_SECRET_NAME`|Name of the AKV secret containing the password for the Apple Installation Certificate.
`APPLE_TEAM_ID_SECRET_NAME`|Name of the AKV secret containing the Apple Developer Team ID.
`APPLE_DEVELOPER_ID_SECRET_NAME`|Name of the AKV secret containing the Apple Developer ID account email address for developer signing.
`APPLE_DEVELOPER_PASSWORD_SECRET_NAME`|Name of the AKV secret containing the Apple Developer ID account password for developer signing.
`APPLE_APPSIGN_ID_SECRET_NAME`|Name of the AKV secret containing the Apple Application Signing ID.
`APPLE_INSTSIGN_ID_SECRET_NAME`|Name of the AKV secret containing the Apple Installation Signing ID.
`LINUX_GPG_PUBLIC_SECRET_NAME`|Name of the AKV secret containing the base64 encoded GPG public key used for Debian package signing.
`LINUX_GPG_PRIVATE_SECRET_NAME`|Name of the AKV secret containing the base64 encoded GPG private key used for Debian package signing.
`LINUX_GPG_PASSPHRASE_SECRET_NAME`|Name of the AKV secret containing the password for the GPG signing key used for the Debian package.
`LINUX_GPG_KEYGRIP_SECRET_NAME`|Name of the AKV secret containing the keygrip of the GPG signing key used for the Debian package.

mjcheetham committed:
fixup! release: create initial Windows installer build workflow

Migrate secrets to Azure Key Vault. Names of the secrets in the AKV, and
the AKV and Managed Identity IDs themselves are stored in GitHub
environment secrets. This indirection allows for an easy way to re-point
these to different Key Vault secrets without modifying the workflow file
itself. In forks, this would allow others to use their own AKV and
sercrets with the same workflow.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! release: add Mac OSX installer build

Continue to migrate secrets to AKV for the macOS build steps.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! release: add signing step for .deb package

Continue to migrate secrets to AKV for the Debian package
build steps.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! build-git-installers: publish gpg public key

Use the `akv-secret` action rather than `az keyvault secret show
| base64 -d` for downloading the Debian package public GPG key.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -101,34 +101,65 @@ jobs:
           mkdir -p .sig &&
-          echo -n "$CODESIGN_P12" | tr % '\n' | base64 -d >.sig/codesign.p12 &&
-          echo -n "$CODESIGN_PASS" >.sig/codesign.pass
+          az keyvault secret download \
```
Do we need to change anything in the setup of the ARM64 runners to ensure that Azure CLI is installed?

mjcheetham replied:
Yep.. and I've just added `az` installation to the image.

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -101,34 +101,65 @@ jobs:
+          az keyvault secret download \
+            --vault-name "$AZURE_VAULT" \
+            --name "$CODESIGN_PASS_SECRET_NAME" \
+            --encoding base64
```
The password is stored in ASCII, isn't it? If so, we probably want to use `show` instead of `download` and `--output tsv` instead of `--encoding base64`, and redirect the output to a file.

mjcheetham replied:
Ah, you are correct that `--encoding base64` is wrong here, but we can still use `az keyvault secret download` rather than `show --output tsv --query value >file`.

> ```
> --encoding -e        : Encoding of the secret. By default, will look for the 'file-encoding' tag
>                        on the secret. Otherwise will assume 'utf-8'.  Allowed values: ascii,
>                        base64, hex, utf-16be, utf-16le, utf-8.
> ```

We can just update the encoding type to `ascii` rather than `base64`. (or should it be `utf-8`?)

dscho replied:
I am not masochist enough to choose a non-ASCII passphrase any time soon, so `ascii` will do 😉

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -101,34 +101,65 @@ jobs:
+            --vault-name "$AZURE_VAULT" \
+            --name "$GPG_PUBLIC_SECRET_NAME" \
+            --query value \
+            --output tsv) &&
```
How can we prevent this from accidentally leaking secrets in case we start debugging via `set -x`? I guess the safest bet (since https://github.com/Azure/get-keyvault-secrets has unfortunately been archived and isn't maintained anymore) would be to add and use a composite Action that is comprised of an `actions/github-script@v7` step that looks somewhat like this:

```yml
- name: obtain KeyVault secret
  id: script
  uses: actions/github-script@v7
  with:
    script: |
      const secret = await exec.exec('az', [
        'keyvault', 'secret', 'show',
        '--vault-name', ${{ toJSON(inputs.azure-vault) }},
        '--name', ${{ toJSON(inputs.secret.name) }},
        '--query', 'value',
        '--output', 'tsv'
      ])
      core.setSecret(secret)
      core.setOutput('result', secret)
```

mjcheetham replied:
We can also use `download` and save to a file, and then `cat` or `<file` the data instead, so it never becomes a bash variable.

dscho replied:
Unfortunately `GPGKEY` is used 28 lines further down, implicitly, by `/usr/src/build-extra/gnupg-with-gpgkey.sh`, which _does_ require it to be a Bash variable. This entire `GPGKEY` business is slightly icky, too, because `GPGKEY` does not _only_ refer to the GPG _key_ ID (which would not need to be secret) but it is actually a partial command-line that includes authorization (continuing how I abused an unquoted `$GPGKEY` in the original version of  Pacman's `repo-add` command, hence fixed).

Therefore, I am afraid that we will need to ensure that the value is marked as a secret in the workflow log...

mjcheetham replied:
I've split up `GPGKEY` into two:

* `WIN_GPG_KEYGRIP_SECRET_NAME`, and
* `WIN_GPG_PASS_SECRET_NAME`

These are both masked, along with all other secrets downloaded via `az keyvault secret show`.

The `GPGKEY` variable is now constructed from the passphrase and keygrip as needed.

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -101,34 +101,65 @@ jobs:
       - name: Build mingw-w64-${{matrix.arch.toolchain}}-git
         env:
-          GPGKEY: "${{secrets.GPGKEY}}"
+          DO_GPG_SIGN: ${{ secrets.WIN_GPG_PUBLIC_SECRET_NAME != '' }}
```
Very nice! It is a change of behavior to test the secret's name instead of the secret's value, but it makes sense, I think.

mjcheetham replied:
Right, we don't actually have the secret value until we try to download it, and if we don't have the _name_ of the secret to download, then we also don't have the value.

dscho reviewed (commented):
Thank you so much! This is already looking great, my only remaining concern is how to ensure that the secret values are masked in the log output (in case future tired me adds `set -x` for debugging, for example). Once this is resolved, I think this is good to go!

mjcheetham committed:
.github/actions/akv-secret: add action to get secrets

Add a new JavaScript GitHub Action to download secrets from Azure Key
Vault using the `az` CLI, mask the secret values, and store them as:
 * outputs,
 * environment variables, or
 * files;

Values are all masked for safe consumption by other steps in a workflow.

Callers of this action can optionally perform base64 decoding of secret
values using the syntax: `INPUT base64> OUTPUT`.

It is assumed that the `az login` command has already been run prior to
this action being invoked.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho reviewed (dismissed):
Very nice!

I could imagine that the resulting workflow definition would be _even_ nicer to read if the convention was introduced that in addition to setting outputs via `NAME => OUTPUT`, secrets could also be downloaded via `NAME ENCODING> FILENAME` (e.g. `${{ secrets.LINUX_GPG_PUBLIC_SECRET_NAME }} base64> msft-git-public.asc` or `${{ secrets.WIN_CODESIGN_PASS_SECRET_NAME }} ascii> .sig/codesign.pass`).

This would probably make even more sense in the context of using the same Action also in other workflow definitions.

What do you think?

dscho commented on .github/actions/akv-secret/index.js:
```diff
@@ -0,0 +1,78 @@
+  } catch (error) {
+    core.setFailed(error.message);
+  }
+})();
```
How about moving this to the end?
```suggestion
})().catch(core.setFailed);
```

dscho commented on .github/actions/akv-secret/index.js:
```diff
@@ -0,0 +1,78 @@
+      if (output.startsWith('$env:')) {
+        // Environment variable
+        const envVarName = output.replace('$env:', '').trim();
+        fs.appendFileSync(process.env.GITHUB_ENV, `${envVarName}=${secretValue}\n`);
```
I believe [this](https://github.com/actions/toolkit/tree/main/packages/core#exporting-variables) is the canonical way to do that:

```suggestion
        core.exportVariable(envVarName, secretValue);
```

That way, all kinds of corner cases are handled correctly, including multi-line values (which require some `<<<EOF` shenanigans).

dscho commented on .github/actions/akv-secret/index.js:
```diff
@@ -0,0 +1,78 @@
+        const fileDir = filePath.substring(0, filePath.lastIndexOf('/'));
+        if (!fs.existsSync(fileDir)) {
+          fs.mkdirSync(fileDir, { recursive: true });
+        }
```
This is a bit fragile in case the `filePath` contains no slash; using [`path.dirname()` ](https://nodejs.org/api/path.html#pathdirnamepath)would be better. And `recursive: true` succeeds if the directory already exists. Maybe something like this?
```suggestion
        fs.mkdirSync(path.dirname(filePath), { recursive: true });
```

dscho reviewed (commented):
I really like where this is going!

dscho commented on .github/actions/akv-secret/action.yml:
```diff
@@ -0,0 +1,54 @@
+        encoded-secret base64> $env:ENV_SECRET
+
+runs:
+  using: node20
```
The commit message still talks about a composite Action. Would you mind changing that, and maybe add a paragraph explaining the woes with the composite Action that forced you to make the code more complex by turning it into a full-fledged JavaScript Action (maybe even linking to that request you found where they asked for dynamic outputs in composite Actions)?

Also, the commit message still talks about the `name => output` form and does not mention that there are now three options: store as output, store as environment variable, store directly to disk. Would you mind adjusting that?

dscho commented on .github/actions/akv-secret/index.js:
```diff
@@ -0,0 +1,135 @@
+const fs = require('fs');
+const os = require('os');
+
+const escapeData = (s) => {
```
This comes from https://github.com/actions/toolkit/blob/%40actions/core%401.1.0/packages/core/src/command.ts#L81-L87, maybe add a comment on top saying so, and also mentioning the deliberate decision not to use `@actions/core` because of the dependency problems that would require `node_modules/` to be committed or, worse, `@vercel/ncc` to be employed?

dscho commented on .github/actions/akv-secret/index.js:
```diff
@@ -0,0 +1,135 @@
+  } catch (error) {
+    setFailed(error);
+  }
+})();
```
I still think that it would be more canonical JavaScript to change that to something like this, but if you prefer it in the current form I won't insist:

```js
(async () => {
  [...]
})().catch(setFailed);
```

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -8,6 +8,10 @@ on:
 permissions:
   id-token: write # required for Azure login via OIDC
 
+env:
+  DO_WIN_CODESIGN: ${{ secrets.WIN_CODESIGN_CERT_SECRET_NAME != '' && secrets.WIN_CODESIGN_PASS_SECRET_NAME != '' }}
+  DO_WIN_GPGSIGN: ${{ secrets.WIN_GPG_KEYGRIP_SECRET_NAME != '' && secrets.WIN_GPG_PRIVATE_SECRET_NAME != '' && secrets.WIN_GPG_PASSPHRASE_SECRET_NAME != '' }}
```
Very, very nice!

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -101,43 +105,62 @@ jobs:
+          vault: ${{ secrets.AZURE_VAULT }}
+          secrets: |
+            ${{ secrets.WIN_CODESIGN_CERT_SECRET_NAME }} base64> home/.sig/codesign.p12
+            ${{ secrets.WIN_CODESIGN_PASS_SECRET_NAME }} base64> home/.sig/codesign.pass
```
I think `codesign.pass` needs the `ascii` encoding, not the `base64` one.

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -195,16 +218,19 @@ jobs:
+          vault: ${{ secrets.AZURE_VAULT }}
+          secrets: |
+            ${{ secrets.WIN_CODESIGN_CERT_SECRET_NAME }} base64> home/.sig/codesign.p12
+            ${{ secrets.WIN_CODESIGN_PASS_SECRET_NAME }} base64> home/.sig/codesign.pass
```
I think this `base64> home/.sig/codesign.pass` needs to be changed to `ascii> ...`, too.

mjcheetham replied:
Omitting any format should be enough here.

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -637,9 +637,9 @@ jobs:
-            ${{ secrets.LINUX_GPG_PASSPHRASE_SECRET_NAME }} > $passphrase
+            ${{ secrets.LINUX_GPG_KEYGRIP_SECRET_NAME }}       > $output:keygrip
+            ${{ secrets.LINUX_GPG_PRIVATE_SECRET_NAME }} base64> $output:private-key
+            ${{ secrets.LINUX_GPG_PASSPHRASE_SECRET_NAME }}    > $output:passphrase
```
This change probably wanted to be squashed into the previous commit ("fixup! release: add signing step for .deb package" instead of "fixup! build-git-installers: publish gpg public key").

dscho commented on .github/workflows/build-git-installers.yml:
```diff
@@ -807,12 +804,12 @@ jobs:
+        with:
+          vault: ${{ secrets.AZURE_VAULT }}
+          secrets: |
+            ${{ secrets.LINUX_GPG_PUBLIC_SECRET_NAME }} base64> deb-package/msft-git-public.asc
```
The commit message still talks about using `az keyvault` directly; Would you mind adjusting it, please?

dscho reviewed (commented):
This looks great! I am a bit embarrassed to ask for yet more changes, but I am confident that this is the last round!

dscho reviewed (approved):
Thank you for indulging me!
